### PR TITLE
Export generated sources

### LINKED ISSUE #8807: Provide access to source code of internal routines to debuggers
Source references already map to internal code - we should provide access to this code, at least the abi encoder / decoder written in yul. This does not need to be part of the metadata because it can be re-generated from the source.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/8807

Open questions:
 - [x] source ID scheme
 - [x] output optimized yul code if optimizer enabled?
 - [ ] CLI support?
 - [x] combined-json support?

### CODE REVIEW

**Comment by chriseth** [comment]
@haltman-at I added the yul ast - please check again (also others, too): https://417293-40892817-gh.circle-artifacts.com/0/soljson.js

**Comment by chriseth** [comment]
We might want to output the optimized yul code if the optimizer is activated.

**Comment by MrChico** [comment]
This would be super nice to have as a `solc --combined-json` option as well 🙏

**Comment by haltman-at** [comment]
OK, I finally had a chance to try this out and it seems to be mostly working but I'm getting some sort of ID collision in certain examples.  Will let you know more details when I have them.

**Comment by haltman-at** [comment]
Oh, nevermind, it's not a real collision.  It's just two different contracts using `utility.yul`.

**Comment by haltman-at** [comment]
Hm, I seem to be occasionally hitting some problem that's resulting in some sort of hang.  No idea what's going on there at the moment, will look into it...

**Comment by haltman-at** [comment]
Nevermind, found the problem, it was just a bug in my code.  Still want to play around with this a little bit more but this seems pretty good to me so far!

**Comment by chriseth** [comment]
@MrChico to be honest, adding this to the commandline will be a little tricky.

File: libsolidity/codegen/CompilerContext.cpp [hunk 1]
```diff
@@ -52,7 +54,6 @@
 // Change to "define" to output all intermediate code
 #undef SOL_OUTPUT_ASM
 #ifdef SOL_OUTPUT_ASM
-#include <libyul/AsmPrinter.h>
 #endif
```
Comment (ekpyron) [inline libsolidity/codegen/CompilerContext.cpp]: Is this empty actually now? Then the ``#ifdef`` should just be removed, shouldn't it?

**Comment by haltman-at** [comment]
OK, having tried this out some more, I'd say this is pretty good!  I've written up a PR to allow Truffle Debugger to make use of it, for whenever it gets released.

**Comment by haltman-at** [comment]
Hm, is there any way that generated sources could not all get the same name?  I don't know, this might not be that important, I think we can handle it if they don't but it would be more convenient if they had different names.

**Comment by chriseth** [comment]
@haltman-at I'm not sure about the name - what would you base the name on? In which context should it be unique?

**Comment by chriseth** [comment]
Documentation and changelog still missing.

**Comment by MrChico** [comment]
> @MrChico to be honest, adding this to the commandline will be a little tricky.

How so?

**Comment by chriseth** [comment]
Tested the optimized output with remix and it works.

**Comment by chriseth** [comment]
Added export via combined-json. Direct commandline export is difficult because this would generated four files per contract.

**Comment by haltman-at** [comment]
> @haltman-at I'm not sure about the name - what would you base the name on? In which context should it be unique?

Well, it would be ideal if sources with different IDs also got assigned different names.  But I'm not sure it's a big deal.  I already changed my Truffle PR for processing this stuff to handle them having the same name, after all...

File: libsolidity/codegen/CompilerContext.cpp [hunk 2]
```diff
@@ -191,14 +190,24 @@ void CompilerContext::appendMissingLowLevelFunctions()
...
 {
-	solAssert(!m_requestedYulFunctionsRan, "requestedYulFunctions called more than once.");
-	m_requestedYulFunctionsRan = true;
+	solAssert(!m_appendYulUtilityFunctionsRan, "requestedYulFunctions called more than once.");
+	m_appendYulUtilityFunctionsRan = true;
 
-	set<string> empty;
-	swap(empty, m_externallyUsedYulFunctions);
-	return {m_yulFunctionCollector.requestedFunctions(), std::move(empty)};
+	std::string code = m_yulFunctionCollector.requestedFunctions();
```
Comment (mijovic) [inline libsolidity/codegen/CompilerContext.cpp]:
```suggestion
	string code = m_yulFunctionCollector.requestedFunctions();
```

File: libsolidity/codegen/CompilerContext.cpp [hunk 3]
```diff
@@ -191,14 +190,24 @@ void CompilerContext::appendMissingLowLevelFunctions()
...
+	if (!code.empty())
+	{
+		appendInlineAssembly(
+			yul::reindent("{\n" + std::move(code) + "\n}"),
```
Comment (mijovic) [inline libsolidity/codegen/CompilerContext.cpp]:
```suggestion
			yul::reindent("{\n" + move(code) + "\n}"),
```

File: libsolidity/codegen/CompilerContext.h [hunk 4]
```diff
@@ -256,7 +258,8 @@ class CompilerContext
 		std::vector<std::string> const& _localVariables = std::vector<std::string>(),
 		std::set<std::string> const& _externallyUsedFunctions = std::set<std::string>(),
 		bool _system = false,
-		OptimiserSettings const& _optimiserSettings = OptimiserSettings::none()
+		OptimiserSettings const& _optimiserSettings = OptimiserSettings::none(),
+		std::string _sourceName = "--CODEGEN--"
```
Comment (mijovic) [inline libsolidity/codegen/CompilerContext.h:265]: Two parameters are missing in docstring, can you add it?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 5]
```diff
@@ -39,6 +39,7 @@
 #include <libyul/Object.h>
 #include <libyul/optimiser/ASTCopier.h>
 #include <libyul/YulString.h>
+#include <libyul/Utilities.h>
```
Comment (mijovic) [inline libsolidity/codegen/ContractCompiler.cpp]: Is this needed?

File: libsolidity/interface/CompilerStack.cpp [hunk 6]
```diff
@@ -591,6 +593,48 @@ evmasm::AssemblyItems const* CompilerStack::runtimeAssemblyItems(string const& _
...
+		_runtime ?
+		c.runtimeGeneratedSources :
+		c.generatedSources;
+	return sources.init([&]{
+		Json::Value sources{Json::arrayValue};
+		// If there is no compiler, then no bytecode was generated and thus no
+		// sources were generated.
+		if (c.compiler)
+		{
+			std::string source =
```
Comment (mijovic) [inline libsolidity/interface/CompilerStack.cpp]:
```suggestion
			string source =
```

File: libsolidity/interface/CompilerStack.cpp [hunk 7]
```diff
@@ -591,6 +593,48 @@ evmasm::AssemblyItems const* CompilerStack::runtimeAssemblyItems(string const& _
...
+				ErrorReporter errorReporter(errors);
+				auto scanner = make_shared<langutil::Scanner>(langutil::CharStream(source, sourceName));
+				yul::EVMDialect const& dialect = yul::EVMDialect::strictAssemblyForEVM(m_evmVersion);
+				shared_ptr<yul::Block> parserResult = yul::Parser{errorReporter, dialect}.parse(scanner, false);
+				solAssert(parserResult, "");
+				sources[0]["ast"] = yul::AsmJsonConverter{sourceIndex}(*parserResult);
+				sources[0]["name"] = sourceName;
+				sources[0]["id"] = sourceIndex;
+				sources[0]["language"] = "Yul";
+				sources[0]["contents"] = std::move(source);
```
Comment (mijovic) [inline libsolidity/interface/CompilerStack.cpp]:
```suggestion
				sources[0]["contents"] = move(source);
```

File: libsolidity/interface/StandardCompiler.cpp [hunk 8]
```diff
@@ -373,6 +380,7 @@ Json::Value collectEVMObject(evmasm::LinkerObject const& _object, string const*
 	output["linkReferences"] = formatLinkReferences(_object.linkReferences);
 	if (_runtimeObject)
 		output["immutableReferences"] = formatImmutableReferences(_object.immutableReferences);
+	output["generatedSources"] = std::move(_generatedSources);
```
Comment (mijovic) [inline libsolidity/interface/StandardCompiler.cpp]:
```suggestion
	output["generatedSources"] = move(_generatedSources);
```

File: libsolidity/interface/StandardCompiler.cpp [hunk 9]
```diff
@@ -1074,25 +1082,27 @@ Json::Value StandardCompiler::compileSolidity(StandardCompiler::InputsAndSetting
 			_inputsAndSettings.outputSelection,
 			file,
 			name,
-			{ "evm.bytecode", "evm.bytecode.object", "evm.bytecode.opcodes", "evm.bytecode.sourceMap", "evm.bytecode.linkReferences" },
+			{ "evm.bytecode", "evm.bytecode.object", "evm.bytecode.opcodes", "evm.bytecode.generatedSources", "evm.bytecode.sourceMap", "evm.bytecode.linkReferences" },
```
Comment (mijovic) [inline libsolidity/interface/StandardCompiler.cpp]: I am not sure if it makes sense to split this to multi lines?

File: libsolidity/interface/StandardCompiler.cpp [hunk 10]
```diff
@@ -1074,25 +1082,27 @@ Json::Value StandardCompiler::compileSolidity(StandardCompiler::InputsAndSetting
...
+				compilerStack.generatedSources(contractName),
 				false
 			);
 
 		if (compilationSuccess && isArtifactRequested(
 			_inputsAndSettings.outputSelection,
 			file,
 			name,
-			{ "evm.deployedBytecode", "evm.deployedBytecode.object", "evm.deployedBytecode.opcodes", "evm.deployedBytecode.sourceMap", "evm.deployedBytecode.linkReferences", "evm.deployedBytecode.immutableReferences" },
+			{ "evm.deployedBytecode", "evm.deployedBytecode.object", "evm.deployedBytecode.opcodes", "evm.deployedBytecode.generatedSources", "evm.deployedBytecode.sourceMap", "evm.deployedBytecode.linkReferences", "evm.deployedBytecode.immutableReferences" },
```
Comment (mijovic) [inline libsolidity/interface/StandardCompiler.cpp]: Same comment for multi lines here

**Comment by chriseth** [comment]
Updated.

File: libsolidity/interface/StandardCompiler.cpp [hunk 11]
```diff
@@ -230,26 +230,29 @@ bool isArtifactRequested(Json::Value const& _outputSelection, string const& _fil
...
-		"evm.deployedBytecode", "evm.deployedBytecode.generatedSources",
-		"evm.deployedBytecode.object", "evm.deployedBytecode.opcodes",
-		"evm.deployedBytecode.sourceMap", "evm.deployedBytecode.linkReferences",
-		"evm.deployedBytecode.immutableReferences",
-		"evm.bytecode", "evm.bytecode.generatedSources",
-		"evm.bytecode.object", "evm.bytecode.opcodes", "evm.bytecode.sourceMap",
-		"evm.bytecode.linkReferences",
 		"evm.gasEstimates", "evm.legacyAssembly", "evm.assembly"
-	};
+	} + evmObjectComponents("bytecode") + evmObjectComponents("deployedBytecode");
```
Comment (mijovic) [inline libsolidity/interface/StandardCompiler.cpp:255]: Looks much better with the helper function :)

**Review by mijovic — APPROVED** [review]
LGTM

**Pull request merged**